Improve C API guidelines for return values (#1129)
python/devguide: developer-workflow/c-api.rst

  fields, arguments and return values are well defined.

- Functions returning ``PyObject *`` must return a valid pointer on success,

>   and ``NULL`` if, and only if, an exception is raised.

Reviewer: Minor syntax suggestion on commas
```suggestion
  and ``NULL`` if and only if an exception is raised.
```

>   Other API must return ``0`` on success,

Reviewer 2: ```suggestion
  Other API must return an ``int``: return ``0`` on success,
```

>   Most other API must return ``0`` on success,

Reviewer 3: Which API is not in this “most” category?
```suggestion
  Other API must return ``0`` on success,
```

Reviewer 4: I think it refers to the ones described below that return e.g. `0` for the lesser return or for "no item found", even though the wording might be a bit better.

>   and a negative value if, and only if, an exception is raised.

Reviewer: ```suggestion
  and a negative value if and only if an exception is raised.
```

>   and a negative value if and only if an exception is raised.
>   Unless there is need to distinguish between error types,
>   ``-1`` should be used as the error value.

Author: @encukou, I suggest either:

```suggestion
  and a ``-1`` if and only if an exception is raised.
```

... or:

```suggestion
  and a negative value if and only if an exception is raised.
```

Author: ```suggestion
  and a negative value if and only if an exception is raised.
  ``-1`` should be used as the error value;
  the caller should use the C idiom ``if (func() < 0)`` to check for errors.
```

Reviewer 2: I prefer to be specific about -1 and suggest testing func() < 0.
